elements/pl-units-input/unit_utils.py

```diff
@@ -0,0 +1,130 @@
+    parsed_atol = ureg.Quantity(atol)
+
+    def magnitude_comparison_fn(
+        submitted_magnitude: float, correct_magnitude: float
+    ) -> bool:
+        """Returns true if submitted_magnitude is close enough to correct_magnitude based on comparison algorithm"""
+        if comparison is ComparisonType.EXACT:
+            return submitted_magnitude == correct_magnitude
+        elif comparison is ComparisonType.SIGFIG:
+            return pl.is_correct_scalar_sf(
+                submitted_magnitude, correct_magnitude, digits
+            )
+        elif comparison is ComparisonType.DECDIG:
+            return pl.is_correct_scalar_dd(
+                submitted_magnitude, correct_magnitude, digits
+            )
+        elif comparison is ComparisonType.RELABS:
+            return pl.is_correct_scalar_ra(
+                submitted_magnitude, correct_magnitude, rtol, parsed_atol.magnitude
+            )
+
+        assert_never(comparison)
+
+    def grade_units_fixed(submitted_ans: str) -> Tuple[float, Optional[str]]:
+        # will return no error, assuming parse() catches all of them
+        parsed_submission = ureg.Quantity(submitted_ans)
+        magnitudes_match = magnitude_comparison_fn(
+            parsed_submission.magnitude, parsed_correct_ans.magnitude
+        )
+        units_match = parsed_correct_ans.units == parsed_submission.units
+
+        if magnitudes_match and units_match:
+            return 1.0, None
+        elif magnitudes_match and not units_match:
+            return 0.7, INCORRECT_UNITS_CORRECT_MAGNITUDE_FEEDBACK
+        elif units_match and not magnitudes_match:
+            return 0.3, CORRECT_UNITS_INCORRECT_MAGNITUDE_FEEDBACK
+
+        return 0.0, INCORRECT_UNITS_AND_MAGNITUDE_FEEDBACK
+
+    return grade_units_fixed
+
+
+def get_units_agnostic_grading_fn(
+    *, ureg: UnitRegistry, correct_ans: str, rtol: float, atol: str
+) -> Callable[[str], Tuple[float, Optional[str]]]:
+    # Assume atol and correct answer have same dimensionality, checked in prepare method
+    correct_ans_base_unit = ureg.Quantity(correct_ans).to_base_units()
+    parsed_atol = ureg.Quantity(atol).to_base_units()
+
+    def grade_units_fixed(submitted_ans: str) -> Tuple[float, Optional[str]]:
+        # will return no error, assuming parse() catches all of them
+        parsed_sub_base_unit = ureg.Quantity(submitted_ans).to_base_units()
+
+        if not correct_ans_base_unit.check(parsed_sub_base_unit.dimensionality):
+            return 0.0, (
+                f"Your answer has dimensionality <code>{parsed_sub_base_unit.dimensionality}</code>, "
+                f"which is inconsistent with <code>{correct_ans_base_unit.dimensionality}</code>."
+            )
+
+        magnitudes_match = pl.is_correct_scalar_ra(
+            parsed_sub_base_unit.magnitude,
+            correct_ans_base_unit.magnitude,
+            rtol,
+            parsed_atol.magnitude,
+        )
+
+        if not magnitudes_match:
```

Comment:
```suggestion
        if magnitudes_match:
            ...
```

Reply:
(and then switch the two return statements)

Reply:
@stevenxpan could you do a multi-line suggest comment? 

c.f. https://stackoverflow.com/a/69044609/1345455